# 01. Importing libraries

In [294]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02 Importing data

In [299]:
# Establish path
path = r'C:\Users\elena\Documents\02.2025 Instacart Basket Analysis'

In [195]:
# Import Orders datafile
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [301]:
# Import Products datafile
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)

In [199]:
# Checking the data imported correctly
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [303]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# 03 Mixed-type data practice exercise

In [204]:
# Creating a small dataframe for mixed-type data exercise
df_test = pd.DataFrame()

In [206]:
# Create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [208]:
# Checking new dataframe
df_test.head()

mix
0     a
1     b
2     1
3  True

In [210]:
# Check for mixed-type column
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\elena\AppData\Local\Temp\ipykernel_28384\1173238358.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [212]:
# Converting mixed-type column to data type string
df_test['mix'] = df_test['mix'].astype('str')

In [214]:
df_test['mix'].dtype

dtype('O')

# 04 Missing values in products data

In [307]:
# Descriptive statistics for products data to get an idea of the variables
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

There are the same number of rows for all variables, suggesting that there are no missing values in these variables. 
However, the statistics are for numerical variables. There might be missing values in string varaibles. 
All values for the variables make sense, so no data problems noted from looking at descriptive statistics.

In [220]:
# Checking for missing values in df_prods dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

Product_name variable has 16 missing values
Product_name is a string variable, so it would not be possible to replace the missing values with mean or median or to extrapolate from the values in the database. 
Next step is to examine the rows with product_name missing, to see if it is possible to determine why the values are missing. 

In [223]:
#Creating a subset from df_prods dataframe which contrains only data for missing values in the product_name column
df_nan = df_prods[df_prods['product_name'].isnull() == True ]

In [225]:
# Checking the new df_nan dataframe
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

There is no obvious reason why the product name values are missing for these rows. Since it is only 16 values, the easiest solution is to create a version of the data with missing values removed and used that version for the analyses.

In [228]:
# Creating dataframe for products file without the units with missing product name.
# Check number of rows in the original file
df_prods.shape

(49693, 5)

In [230]:
# Creating dataframe with no missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [232]:
# Checking rows in the new df_prods_clean dataframe
df_prods_clean.shape

(49677, 5)

The 16 rows with missing products names are not present in the new dataframe. 

# 05 Outliers in the products data

In [311]:
# Descriptive statistics for the cleaned products data 
df_prods_clean.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The max value in the 'prices' column is much higher than the upper quartile or the median. Also the mean is higher than the median. There might be several outliers on the higher end of the 'prices' distribution and it is influencing the mean. 

In [330]:
# Calculate IQR
Q1=4.1
Q3=11.1
IQR = Q3 - Q1
IQR

7.0

In [328]:
# Define outliers as values outside 1.5 * IQR from Q1 and Q3
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

In [340]:
lower_bound

-6.4

In [342]:
upper_bound

21.6

In [334]:
# Identify outliers
outliers = df_prods_clean[(df_prods_clean['prices'] < lower_bound) | (df_prods_clean['prices'] > upper_bound)]

In [338]:
# Determine how many outliers there are in the 'prices' column
outliers

product_id                                       product_name  \
39             40  Beef Hot Links Beef Smoked Sausage With Chile ...   
83             84                                         Lamb Shank   
1932         1932                             Soy Chorizo Vegetarian   
1992         1992                       Beef (101445) Summer Sausage   
2005         2005                         Fresh Ground Turkey Breast   
...           ...                                                ...   
48343       48339                                      Lobster Tails   
48592       48588                         All Natural Ground Chicken   
48784       48780                              Olsen Pickled Herring   
48966       48962                       Hardwood Smoked Sliced Bacon   
49440       49436                              Imitation Crab Flakes   

       aisle_id  department_id  prices  
39          106             12    22.5  
83            7             12    24.3  
1932        106             12    23.8  
1992        106             12    24.5  
2005         35             12    22.5  
...         ...            ...     ...  
48343        39             12    24.9  
48592        49             12    21.7  
48784        39             12    22.2  
48966       106             12    23.1  
49440        15             12    23.5  

[191 rows x 5 columns]

There are 191 rows with prices higher than the upper_bound, so technically these are outliers. However, the distribution is right skewed and the prices a little above the upper_bount of 21.6 do not seem unreasonable in the context of the data. It needs to be determine if there are any outliers that are unreasonaly large and are likely to be a mistake. Next step is to identify top 5 values in the 'prices' column.

In [346]:
# Find the top 5 highest values in column 'prices'
top_5_prices = df_prods_clean['prices'].nlargest(5)

In [348]:
# Display top 5 prices
print(top_5_prices)

33666    99999.0
21554    14900.0
9020        25.0
19392       25.0
21468       25.0
Name: prices, dtype: float64


Price of 25 seems reasonable in context of the data, while values 99999 and 14900 seen very improbable. These might be a mistake or might have been entered in place of missing values. Since it is only two values, it would be most reasonable to exclude the two rows of data from the dataframe, so they do not interfere with any analyses of the data. 

In [351]:
# Create a new dataframe with values in column 'prices' lower than 1000
df_prods_clean2 = df_prods_clean[df_prods_clean['prices'] < 1000]

In [353]:
# Check the new dataframe
df_prods_clean2.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [355]:
# Descriptive statistics for the new dataframe
df_prods_clean2.describe()

product_id      aisle_id  department_id        prices
count  49675.000000  49675.000000   49675.000000  49675.000000
mean   24850.083181     67.761973      11.728515      7.680552
std    14340.815137     38.316673       5.850706      4.199287
min        1.000000      1.000000       1.000000      1.000000
25%    12432.500000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.500000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

Now mean for the prices is close to the median and the max value looks realistic. 

In [360]:
# Check the shape of the new dataframe to see how many rows it now has
df_prods_clean2.shape

(49675, 5)

# 06 Duplicate values in products data

In [365]:
# Creating a separate dataframe for duplicate values
df_dups = df_prods_clean2[df_prods_clean2.duplicated()]

In [367]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

df_prods_clean2 dataframe contains 49675 rows

In [370]:
# Dropping duplicates from the dataframe
df_prods_clean2_no_dups = df_prods_clean2.drop_duplicates()

In [372]:
# Checking number of row in the new dataframe without duplicates
df_prods_clean2_no_dups.shape

(49670, 5)

5 duplicates are not present in the df_prods_clean_no_dups dataframe

# 07 Exporting the checked products dataframe

In [376]:
# Exporting the df_prods_clean_no_dups dataframe as products_checked csv file
df_prods_clean2_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [378]:
# End of Exercise 4.5.1

# 08 Exercise 4.5.2-4.5.9 - Data checking orders data

In [252]:
# Exercise 4.5.2 - orders data preliminary check
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Descriptive statistics show:
- There are fewer values in the "days_since_prior_order" than all the other columns, suggesting some missing values in this column
- Min an max values for "order_day_of_week is consistent with 7 day week
- Min an max values for "order_hour_of_day is consistent with 24 hour day
- Descriptive statistics do not show any unreasonable values for the variables

In [255]:
# Exercise 4.5.3 - checking for mixed-type variables in orders data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len(df_test[weird]) > 0:
    print(col)

C:\Users\elena\AppData\Local\Temp\ipykernel_28384\3235787353.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len(df_test[weird]) > 0:
C:\Users\elena\AppData\Local\Temp\ipykernel_28384\3235787353.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len(df_test[weird]) > 0:
C:\Users\elena\AppData\Local\Temp\ipykernel_28384\3235787353.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len(df_test[weird]) > 0:
C:\Users\elena\AppData\Local\Temp\ipykernel_28384\3235787353.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len(df_test[weird]) > 0:
C:\Users\elena\AppData\Local\Temp\ipykernel_28384\3235787353.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len(df_test[weird]) > 0:
C:\Users\elena\AppData\Local\Temp\ipykernel_28384\3235787353.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame

No mixed-type columns were found 

In [258]:
# Exercise 4.5.4 - confirm data type in each column in orders data
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


In [260]:
# Exercise 4.5.5 - checking for missing values in orders data
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The variable that has missing values is the 'days_since_prior_order' which has 206209 values missing.
This needs to be investigated further by looking at the subset of data where days_since_prior_order is missing.

In [263]:
# Creating a data frame for rows with missing days_since_prior_order
df_ords_miss = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [265]:
df_ords_miss.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
11          11   2168274        2             1                   2   
26          26   1374495        3             1                   1   
39          39   3343014        4             1                   6   
45          45   2717275        5             1                   3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

Looking a the head of the df_ords_miss, days_since_prior_order are missing for first orders.

In [268]:
# Checking that all the orders with days_since_prior_order missing are first orders (order_number=1)
count = (df_ords_miss['order_number'] > 1).sum()

There are no orders with order_number > 1 in the df_ords_miss dataframe. It makes sense that there is no days since prior order for the first order.  

In [271]:
# Exercise 4.5.6 - dealing with missing values in orders data

The missing values are in the 'days_since_prior_order' column, specifically there are no values in this column for the first orders (order_number=1), which is appropriate since there have not been prior orders at that point. Therefore, there is no need to do anything about the missing values. 

In [274]:
# Exercise 4.5.7 and 4.5.8 -checking for duplicates and dealing with duplicates in orders data
df_ords_dups = df_ords[df_ords.duplicated()]

In [275]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates found in the df_ords dataframe. Therefore, no action for dealing with duplicates needed. 

# 09 Dropping a redundant column from orders dataframe

In [283]:
# There is 'Unnamed: 0' column in the orders dataframe that is redundant. 
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [285]:
# Dropping the 'Unnamed: 0' column and creating cleaned dataframe
df_ords_clean = df_ords.drop('Unnamed: 0', axis=1)

In [287]:
# Checling cleaned dataframe
df_ords_clean.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 10 Exporting the checked orders dataframe

In [289]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))